### PR TITLE
teach souper about some more data flow analysis like nonZero, nonNega…

### PR DESCRIPTION
**Pull request by jubitaneja**
…tive and powerOfTwo

### CODE REVIEW

File: lib/Extractor/KLEEBuilder.cpp [hunk 1]
```diff
@@ -143,14 +149,26 @@ ref<Expr> ExprBuilder::makeSizedArrayRead(unsigned Width, StringRef Name,
   std::vector<unsigned> ZeroBits, OneBits;
   UpdateList UL(Arrays.back().get(), 0);
   ref<Expr> Var = ReadExpr::create(UL, klee::ConstantExpr::alloc(0, Expr::Int32));
-  if (Origin && Origin->K == Inst::Var && (Origin->KnownZeros.getBoolValue() ||
-                                           Origin->KnownOnes.getBoolValue())) {
```
Comment (regehr) [inline lib/Extractor/KLEEBuilder.cpp]: this'll look better if you put the whole or-expression on the second line

File: lib/Extractor/Candidates.cpp [hunk 2]
```diff
@@ -45,6 +45,10 @@ static llvm::cl::opt<bool> HarvestKnownBits(
     "souper-harvest-known-bits",
     llvm::cl::desc("Perform known bits analysis (default=true)"),
     llvm::cl::init(true));
+static llvm::cl::opt<bool> HarvestMoreKnownBits(
+    "souper-harvest-more-known-bits",
+    llvm::cl::desc("Perform more known bits analysis to find if non-zero(default=false)"),
+    llvm::cl::init(false));
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: Please just put all known bits until control of a single flag

File: lib/Extractor/KLEEBuilder.cpp (hunk 1 above)
Comment (jubitaneja) [inline lib/Extractor/KLEEBuilder.cpp]: I think you meant the OR expression to look like as its in line : 153 below. The line is deleted for which you pointed a change. (?)

File: lib/Extractor/KLEEBuilder.cpp [hunk 3]
```diff
@@ -143,14 +149,26 @@ ref<Expr> ExprBuilder::makeSizedArrayRead(unsigned Width, StringRef Name,
...
-    ref<Expr> NotZeros = NotExpr::create(klee::ConstantExpr::alloc(Origin->KnownZeros));
-    ref<Expr> VarOrNotZero = OrExpr::create(Var, NotZeros);
-    ZeroBitsMap[Origin] = EqExpr::create(VarOrNotZero, NotZeros);
-    ref<Expr> Ones = klee::ConstantExpr::alloc(Origin->KnownOnes);
-    ref<Expr> VarAndOnes = AndExpr::create(Var, Ones);
-    OneBitsMap[Origin] = EqExpr::create(VarAndOnes, Ones);
+  if (Origin && Origin->K == Inst::Var) {
+    if (Origin->KnownZeros.getBoolValue() || Origin->KnownOnes.getBoolValue()) {
```
Comment (jubitaneja) [inline lib/Extractor/KLEEBuilder.cpp]: Does this look ok?

**Comment by regehr** [comment]
Here's a case where we need a better error message, or better yet just support the flags in either order.

```
pc33:souper-jubi regehr$ cat more-knownbits1.opt
%0:i32 = var
%1:i32 = var (z) (00000000000000000000000000000000)
%2:i32 = shlnw %0, %1
%3:i32 = and %2, 2147483648:i32
%4:i1 = eq %3, 0:i32
infer %4
pc33:souper-jubi regehr$ ./build/souper-check $SOUPER_SOLVER -infer-rhs -print-replacement < more-knownbits1.opt 
<stdin>:2:19: expected inst, block, cand, infer, result, pc, or blockpc
pc33:souper-jubi regehr$ 
```

File: test/Tool/more-knownbits11.opt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+; RUN: %souper-check %solver -print-counterexample=false  %s > %t 2>&1
+; RUN: FileCheck %s < %t
+
+; CHECK: LGTM
+%0:i8 = var (p)
+%1:i8 = var (z)
+%2:i8 = udivexact %0, %1
+%3:i8 = ctpop %2
+%4:i1 = eq %3, 1:i8
+cand %4 1:i1
```
Comment (regehr) [inline test/Tool/more-knownbits11.opt]: this one might not be testing what you think it is testing, it is still LGTM when the (z) is removed

File: test/Tool/more-knownbits5.opt [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+; CHECK: %0:i16 = var (p)
+; CHECK-NEXT: %1:i16 = var (z)
+; CHECK-NEXT: %2:i16 = udivexact %0, %1
+; CHECK-NEXT: %3:i16 = ctpop %2
+; CHECK-NEXT: %4:i1 = eq 1:i16, %3
+; CHECK-NEXT: cand %4 1:i1
+
+%0:i16 = var (p)
+%1:i16 = var (z)
```
Comment (regehr) [inline test/Tool/more-knownbits5.opt]: this one works even without the (z)

File: test/Tool/more-knownbits11.opt (hunk 4 above)
Comment (jubitaneja) [inline test/Tool/more-knownbits11.opt]: I added the 'z' flag just to make sure that denominator of division operation is never zero. The intention was basically to test power of two feature only. I have fixed the test case by removing 'z' flag. Similarly in other test case too.

**Comment by regehr** [comment]
I'm getting an apparent infinite loop in parser_tests. It just hangs up, both in a Debug and Release build.

**Pull request merged**